### PR TITLE
[dynamo] Return original tensor for torch in-place ops

### LINKED ISSUE #113271: [dynamo] Assigning result of Tensor in-place op destroys mutation tracking
### 🐛 Describe the bug

The new tensor variable returned by e.g.

```python
x = x.add_(1)
```

Does not retain mutation tracking, similar to builtin in-place magic methods https://github.com/pytorch/pytorch/issues/113160

Repro incoming

### Versions

main

cc @ezyang @bdhirsh @anijain2305 @chauhang @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @msaroufim @wconstab @zou3519 @aakhundov

### PR DESCRIPTION
**Pull request by jon-chuang**
Fixes https://github.com/pytorch/pytorch/issues/113271

Similar to https://github.com/pytorch/pytorch/pull/113171

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

### CODE REVIEW

File: torch/_dynamo/variables/tensor.py [hunk 1]
```diff
@@ -788,6 +791,13 @@ def _register_hook_trampoline(tensor):
                 ),
                 **options,
             )
+            if is_inplace_op:
+                assert (
+                    ret.as_proxy().node.meta["example_value"]
```
Comment (jon-chuang) [inline torch/_dynamo/variables/tensor.py:796]: Assert may be too strong. Running branch if the example_values are identical may be the solution if tests fail.

Comment (jon-chuang) [reply torch/_dynamo/variables/tensor.py:802]: Had to do this. Some non-inplace ops collide into the name condition.

File: torch/_dynamo/variables/tensor.py [hunk 2]
```diff
@@ -788,6 +789,21 @@ def _register_hook_trampoline(tensor):
                 ),
                 **options,
             )
+            if (
+                not name.startswith("_")
```
Comment (jon-chuang) [inline torch/_dynamo/variables/tensor.py:793]: do we even need to check this? As long as tensor method returns itself as example value, we should probably return the original tensor regardless?

File: torch/_dynamo/variables/torch.py [hunk 3]
```diff
@@ -750,6 +750,16 @@ def check_any_unspec(x):
                 else:
                     unimplemented(f"out variant of {type(kwargs['out'])}")
 
+            name = fn_.__name__
+            if (
+                not name.startswith("_")
+                and name.endswith("_")
```
Comment (ezyang) [inline torch/_dynamo/variables/torch.py:756]: This is not a good way to test this. I think we should infer this by testing the fake tensor output when we fake prop to see if tensor identity was preserved.

**Review by jansel — CHANGES_REQUESTED** [review]
I think we also need to set `get_mutation_names()` on the created IR node.  This is needed so inductor doesn't reorder the kernel with respect to other users of args[0].

**Comment by jon-chuang** [comment]
I think might not be able to do this for `set_`, even though technically the returned tensor is identical to previous... that's because the metadata (e.g. shape) might change?

**Comment by ezyang** [comment]
I'm confused; do you want to propose https://github.com/pytorch/pytorch/pull/113171 or this PR as the one to actually land?

**Comment by jon-chuang** [comment]
We'll see if the fix discussed for #113171 will fix this automatically - presumably yes. 

I'll set this to draft for now as we're taking a less invasive/more general approach there.

**Comment by jansel** [comment]
> @jansel would this also affect the proposed more general fix here: [#113171 (comment)](https://github.com/pytorch/pytorch/pull/113171#discussion_r1387211982)?
> 
> I'm not sure I understand what VTs have to do with reordering of inductor IR. Why isn't the ordering handled in the output graph fx trace?

Aliasing and mutation needs to be correctly exposed on the inductor.ir nodes.  This would be a correctness issue.

File: test/dynamo/test_subclasses.py [hunk 4]
```diff
@@ -504,8 +504,8 @@ def forward(self, L_x_ : torch.Tensor):
 
     class GraphModule(torch.nn.Module):
         def forward(self, l_x_):
-            add_ = l_x_.add_(1.0);  l_x_ = None
-            return (add_,)
+            add_ = l_x_.add_(1.0)
+            return (l_x_,)
```
Comment (jon-chuang) [inline test/dynamo/test_subclasses.py:508]: @jansel regarding the aliasing information in inductor, are you referring to the effect seen above this PR has on the fx trace?

Comment (jon-chuang) [reply test/dynamo/test_subclasses.py:508]: @bdhirsh will this change adversely affect how AOTAutograd functionalizes mutations?

Comment (ezyang) [reply test/dynamo/test_subclasses.py:508]: No, this won't matter at all; post functionalization you'll get the same IR for both before and after

**Comment by jansel** [comment]
By aliasing I mean stuff like view ops.  View ops can't be extern kernels, because inductor may think the memory is free if it doesn't know the input aliases the output.

**Comment by jon-chuang** [comment]
Hmm are there any view ops that are actually in-place? Doesn't view always return a new tensor?

**Comment by ezyang** [comment]
I think you and jansel are talking past each other. What is going on with `add_` and friends is not really aliasing at all, it's just having multiple references to the same Tensor.

**Pull request closed without merging**